app/assets/javascripts/toc.js

```diff
@@ -1,28 +1,88 @@
-$(document).ready(function() {
-  let element = $('#toc')
-  if(element.length > 0) {
-    let selectedId = element.data('selected')
-    let baseUrl = element.data('url')
-    let initialUrl = `${baseUrl}?node=${selectedId}&full=true`
-
-    $.getJSON(
-      initialUrl,
-      function(data) {
-        element.tree({
-          data: data
-        });
-      }).done(function() {
-        let selectedNode = element.tree('getNodeById', selectedId);
-        element.tree('selectNode', selectedNode);
-    });
+class TocBuilder {
+  constructor(element) {
+    this.element = $(element)
+    // We need a separate data element that is updated by turbolinks so we always
+    // have the correct selected node id. The toc element itself is permanent (not
+    // updated by turbolinks), and is instead updated by event-triggered javascript.
+    this.dataElement = $(`${element}-data`)
+  }
+
+  get baseUrl() {
+    return '/toc/'
+  }
+
+  get initialUrl() {
+    return `${this.baseUrl}?node=${this.selectedId}&full=true`
+  }
+
+  get selectedId() {
+    return this.dataElement.data('selected')
+  }
+
+  build() {
+    this.setupTree()
+    this.setupEventHandlers()
+  }
 
-    $(element).on(
-      'tree.click',
-      function(event) {
-          let node = event.node
-          let url = `/catalog/${node.id}`
-          window.location.href = url
+  setupEventHandlers() {
+    // Listen for click on the leaf node and follow link to component
+    this.element.on('activate_node.jstree', function (e, data) {
+      let location = `/catalog/${data.node.id}`
+      Turbolinks.visit(location)
+      // Scroll to top of page instead of jumping for better user experience
+      window.scrollTo({ top: 0, behavior: 'smooth' });
```

Comment:
Question: Is there a way to make it scroll up after the page loads? I don't know if it'll be like this in prod, but I run into this issue where it scrolls up to old content and just looks like it's broken.

Reply:
You mean it scrolls before the refreshed content has finished loading?

Reply:
Yeah.

Reply:
Getting smooth scroll to play nice with turbolinks rendering is a challenge. I'm going to remove it and just default back to standard behavior for now. I think it looks fine without it. We can revisit at a later date if it becomes an issue.